### PR TITLE
GEODE-10039: Fix the issue that BucketProfile can be stale in rare cases.

### PR DESCRIPTION
**Pull request by jchen21**
### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

**Comment by DonalEvans** [comment]
Does this PR replace [this one](https://github.com/apache/geode/pull/7364) that's still open? Both are for the same ticket and have the same title. Is this PR different in approach to the previous one, and if so, should [the comment on the JIRA ticket](https://issues.apache.org/jira/browse/GEODE-10039?focusedCommentId=17500436&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17500436) describing the proposed solution be amended?

**Comment by kirklund** [comment]
How about "Prevent stale BucketProfiles for queues" for a commit message?

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+import org.apache.geode.internal.cache.partitioned.RemoveBucketMessage;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+import org.apache.geode.test.dunit.AsyncInvocation;
+import org.apache.geode.test.dunit.rules.ClientVM;
+import org.apache.geode.test.dunit.rules.ClusterStartupRule;
+import org.apache.geode.test.dunit.rules.DistributedBlackboard;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.junit.rules.serializable.SerializableTestName;
+
+public class RebalanceWhileCreatingRegionDistributedTest implements Serializable {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:45]: Test looks ok. My personal preference is to write this with `DistributedReference` rule instances for each VM and use `AtomicReference<CountDownLatch>` instead of the blackboard rule. Seems ok though.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            (DestroyRegionOperation.DestroyRegionMessage) message;
+        if (drm.regionPath.contains(regionName)) {
+          logger.info(
+              "TestDistributionMessageObserver.beforeSendMessage about to wait for After_CreateProxyRegion gate regionName={}",
+              drm.regionPath);
+          try {
+            blackboard.waitForGate(AFTER_CREATE_PROXY_REGION);
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:290]: In DUnit tests, you can usually remove the try-catch blocks and just keep adding throws checked exceptions to every method.

File: geode-core/src/main/java/org/apache/geode/internal/cache/DestroyRegionOperation.java [hunk 3]
```diff
@@ -167,32 +167,23 @@ public void run() {
...
+                    .getPartitionedRegionUsingBucketRegionName(dm.getCache(), regionPath);
+                String bucketName = PartitionedRegionHelper.getBucketName(regionPath);
+                if (partitionedRegion != null && bucketName != null) {
+                  int bucketId = PartitionedRegionHelper.getBucketId(bucketName);
+                  partitionedRegion.getRegionAdvisor().removeIdAndBucket(bucketId, getSender(),
+                      serialNum, op.isRegionDestroy() && !op.isClose());
+                }
               } catch (RegionDestroyedException ignore) {
-                // ditto
+                // region not found - it's been destroyed
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/DestroyRegionOperation.java:180]: How about `// nothing to do because Region is already destroyed`?

File: geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegionHelper.java [hunk 4]
```diff
@@ -717,6 +714,31 @@ public static Bucket getProxyBucketRegion(Cache cache, String fullPath, boolean
...
+      return null;
+    }
+    String prid = getPRPath(bucketName);
+    Region region;
+    final InitializationLevel oldLevel = LocalRegion.setThreadInitLevelRequirement(ANY_INIT);
+    try {
+      region = cache.getRegion(prid);
+    } finally {
+      LocalRegion.setThreadInitLevelRequirement(oldLevel);
+    }
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegionHelper.java:735]: Perfect!

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java [hunk 5]
```diff
@@ -180,19 +180,23 @@ public void processProfilesQueuedDuringInitialization() {
...
+              buckets[qbp.bucketId].getBucketAdvisor().removeIdWithSerial(qbp.memberId,
+                  qbp.serials[0], qbp.destroyed);
+            } else {
+              for (int i = 0; i < buckets.length; i++) {
+                BucketAdvisor ba = buckets[i].getBucketAdvisor();
+                int serial = qbp.serials[i];
+                if (serial != ILLEGAL_SERIAL) {
+                  ba.removeIdWithSerial(qbp.memberId, serial, qbp.destroyed);
+                }
+              } // for
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:193]: Good place to extract this as a private method of its own. The purpose isn't to allow multiple code paths to call it, just to keep methods shorter for JIT optimization and make it more readable (if you use a good method name).

File: geode-core/src/test/java/org/apache/geode/internal/cache/DestroyRegionOperationJUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+import org.junit.Test;
+
+import org.apache.geode.cache.Operation;
+import org.apache.geode.distributed.DistributedSystem;
+import org.apache.geode.distributed.internal.membership.InternalDistributedMember;
+
+public class DestroyRegionOperationJUnitTest {
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/internal/cache/DestroyRegionOperationJUnitTest.java:30]: Since this is a new test class, please just name it `DestroyRegionOperationTest`. The wiki pages on writing tests say to name them "FooTest, FooIntegrationTest, FooDistributedTest", etc.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    ClientVM client2 =
+        cluster.startClientVM(6, c -> c.withServerConnection(accessor.getPort())
+            .withCacheSetup(cf -> cf.setPoolReadTimeout(20000)));
+
+    // Do puts
+    client2.invoke(() -> {
+      Region<Integer, Integer> region =
+          ClusterStartupRule.clientCacheRule.createProxyRegion(regionName);
+      IntStream.range(0, 3).forEach(i -> region.put(i, i));
+    });
```
Comment (agingade) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:125]: It will be good to add validation for some of the expected state. 
- Wait for accessor to finish and see it proxy-region creation completed successfully. Accessing that region in accessor node to fetch and perform cache operation (say put and get) to see there is no issue with the region and all the expected data is present.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+      ClusterStartupRule.getCache().close();
+    });
+
+    asyncInvocation.get();
+
+    // Make sure the accessor knows that the primary bucket has moved to server2
+    accessor.invoke(() -> {
+      PartitionedRegion pr =
+          (PartitionedRegion) ClusterStartupRule.getCache().getRegion(regionName);
+      assertThat(pr.getRegionAdvisor().getBucket(0).getBucketAdvisor().getProfile(source))
```
Comment (agingade) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:193]: Adding check to see if the bucket is moved to destination will confirm the bucket has moved successfully

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java (hunk 5 above)
Comment (agingade) [reply geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:193]: Please create a new ticket to refactor the QueuedBucketProfile.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java [hunk 9]
```diff
@@ -2544,4 +2544,9 @@ void superTxApplyPut(Operation putOp, Object key, Object wrappedNewValue, boolea
         versionTag, tailKey);
   }
 
+  @Override
+  Set<InternalDistributedMember> getDestroyRegionRecipients() {
+    return getSystem().getDistributionManager().getOtherDistributionManagerIds();
```
Comment (agingade) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java:2549]: Can we add comment to say why all distributed members are fetched...

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+  public DistributedBlackboard blackboard = new DistributedBlackboard();
+
+  private static final Logger logger = LogService.getLogger();
+
+  public static final String BEFORE_REMOVE_BUCKET_MESSAGE = "Before_RemoveBucketMessage";
+
+  public static final String AFTER_CREATE_PROXY_REGION = "After_CreateProxyRegion";
+
+  @Test
+  public void testRebalanceDuringRegionCreation() throws Exception {
```
Comment (gesterzhou) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:63]: tests are good. However, it will be a little bit difficult for people with less background to understand the design of the test. 

It's better to add some comments. 

There're following key points: 
- simulate a bucket moving scenario
- When processing RemoveBucketMessage, it will calculate recipients first, at that moment another member (proxy) might have not initialized PR yet.
- send DestryRegionOperation.DestroyRegionMessage to let other member to remove the profile, but the proxy will miss it.

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java [hunk 11]
```diff
@@ -180,19 +180,23 @@ public void processProfilesQueuedDuringInitialization() {
...
                   "applying queued profile removal for all buckets for {}", qbp.memberId);
             }
-            for (int i = 0; i < buckets.length; i++) {
-              BucketAdvisor ba = buckets[i].getBucketAdvisor();
-              int serial = qbp.serials[i];
-              if (serial != ILLEGAL_SERIAL) {
-                ba.removeIdWithSerial(qbp.memberId, serial, qbp.destroyed);
-              }
-            } // for
+            if (qbp.serials.length == 1) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:183]: It is not clear to me why this change was made. It seems like the old code required that the the "serials" array length was >= to the "buckets.length". But now you allow "serials.length" to be 1 and in that case do not check for ILLEGAL_SERIAL. What corresponding change required this one?

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java [hunk 12]
```diff
@@ -426,6 +430,23 @@ public void removeIdAndBuckets(InternalDistributedMember memberId, int prSerial,
...
+      if (preInitQueue != null) {
+        // Queue profile during pre-initialization
+        QueuedBucketProfile qbf =
+            new QueuedBucketProfile(bucketId, memberId, serial, regionDestroyed);
+        preInitQueue.add(qbf);
+        return;
+      }
+    }
+
+    if (buckets != null) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:445]: Is it really okay to just do nothing here if buckets is null? It does not get nulled out at the end of its life. It is null after the constructor and made non-null when initializeRegionAdvisor is called. But it is called BEFORE we call processProfilesQueuedDuringInitialization which is the only place we null out preInitQueue. We only get to this null check if preInitQueue is null. My concern is if we do get to this line and buckets is null then something is wrong with our internal logic and we are not going to  call removeIdWithSerial which seems wrong.

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java (hunk 11 above)
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:183]: it looks the "if" can be handled in "else"

Comment (jchen21) [reply geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:183]: Looking at the code again, when `serials.length == 1`, the `bucketId` might not necessarily be `0`. Therefore, we still need the `if` and `else`.

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java (hunk 12 above)
Comment (jchen21) [reply geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:445]: If `buckets` is `null`, which means `initializeRegionAdvisor()` and `processProfilesQueuedDuringInitialization()` are not called yet. At this time, as long as `RegionAdvisor` constructor is called, `preInitQueue` is not `null`. It will enqueue the bucket profile and `return` in the `if (preInitQueue != null)` block line 441.

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java (hunk 11 above)
Comment (jchen21) [reply geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:183]: The corresponding change is in `removeIdAndBucket()` which enqueues a `QueuedBucketProfile` that has a single serial to destroy. I missed `ILLEGAL_SERIAL` checking here. The serials array is used to for `DestroyPartitionedRegionMessage`. Here we use the single element serials array for `DestroyRegionMessage`.

**Review by kirklund — APPROVED** [review]

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java (hunk 7 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:126]: Additional verification added.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java (hunk 8 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:202]: Additional check added.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java (hunk 2 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:301]: This method overrides `DistributionMessageObserver.beforeSendMessage()` which does not have throws declaration.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java (hunk 9 above)
Comment (jchen21) [reply geode-core/src/main/java/org/apache/geode/internal/cache/BucketRegion.java:2553]: Comments added.

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java (hunk 12 above)
Comment (jchen21) [reply geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/RegionAdvisor.java:445]: If we do get to this line, then `preInitQueue` has to be `null`, which means `processProfilesQueuedDuringInitialization()` is called. And `initializeRegionAdvisor()` is called before `processProfilesQueuedDuringInitialization()` to initialize the `buckets` array. So I don't think `buckets` is `null`. Maybe I am missing something?

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java (hunk 10 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java:63]: Comments added.

**Comment by jchen21** [comment]
> Does this PR replace [this one](https://github.com/apache/geode/pull/7364) that's still open? Both are for the same ticket and have the same title. Is this PR different in approach to the previous one, and if so, should [the comment on the JIRA ticket](https://issues.apache.org/jira/browse/GEODE-10039?focusedCommentId=17500436&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17500436) describing the proposed solution be amended?

@DonalEvans Good point. I have closed the outdated pull request and updated the JIRA with some more comments.

File: geode-core/src/main/java/org/apache/geode/internal/cache/DestroyRegionOperation.java [hunk 13]
```diff
@@ -167,32 +167,23 @@ public void run() {
...
-              CacheDistributionAdvisee advisee = null;
               try {
-                advisee = PartitionedRegionHelper.getProxyBucketRegion(dm.getCache(), regionPath,
-                    waitForBucketInitializationToComplete);
-              } catch (PRLocallyDestroyedException ignore) {
-                // region not found - it's been destroyed
+                PartitionedRegion partitionedRegion = PartitionedRegionHelper
+                    .getPartitionedRegionUsingBucketRegionName(dm.getCache(), regionPath);
+                String bucketName = PartitionedRegionHelper.getBucketName(regionPath);
+                if (partitionedRegion != null && bucketName != null) {
```
[the remark this answers is not in the document]
Comment (jchen21) [reply geode-core/src/main/java/org/apache/geode/internal/cache/DestroyRegionOperation.java]: This is done.

**Review by dschneider-pivotal — APPROVED** [review]
